packages/protocol/contracts/L1/libs/LibProposing.sol

```diff
@@ -74,35 +75,47 @@ library LibProposing {
             revert L1_TOO_MANY_BLOCKS();
         }
 
-        bytes32 parentMetaHash =
-            _state.blocks[(b.numBlocks - 1) % _config.blockRingBufferSize].metaHash;
-        // assert(parentMetaHash != 0);
+        TaikoData.Block storage parentBlock =
+            _state.blocks[(b.numBlocks - 1) % _config.blockRingBufferSize];
 
         // Check if parent block has the right meta hash. This is to allow the proposer to make sure
         // the block builds on the expected latest chain state.
-        if (params.parentMetaHash != 0 && parentMetaHash != params.parentMetaHash) {
+        if (params.parentMetaHash != 0 && parentBlock.metaHash != params.parentMetaHash) {
             revert L1_UNEXPECTED_PARENT();
         }
 
+        // Verify the passed in L1 state block number.
+        // We only allow the L1 block to be 2 epochs old.
+        // The other constraint is that the L1 block number needs to be larger than or equal the one
+        // in the previous block.
+        if (
+            params.l1StateBlockNumber < block.number - 64
+                || params.l1StateBlockNumber >= block.number
+                || params.l1StateBlockNumber < parentBlock.l1StateBlockNumber
+        ) {
+            revert L1_INVALID_L1_STATE_BLOCK();
+        }
+
         // Initialize metadata to compute a metaHash, which forms a part of
         // the block data to be stored on-chain for future integrity checks.
         // If we choose to persist all data fields in the metadata, it will
         // require additional storage slots.
         unchecked {
             meta_ = TaikoData.BlockMetadata({
-                l1Hash: blockhash(block.number - 1),
+                l1Hash: blockhash(params.l1StateBlockNumber),
```

Comment:
I think even if we keep `l1Hash: blockhash(block.number - 1)` it should still work, right?

Reply:
Same question for `timestamp`, `l1Height` and `difficulty` below.  If in Raiko we need to access `l1StateBlockNumber`, we can still keep`timetamp` and `difficulty` as-is and only change `l1Height`. 12 seconds per L1 block is not accurate.

Reply:
The thing necessary for preconfirmations is that the whole L2 block is fully deterministic within a reasonable window to give some extra flexibility to when the block can be proposed without changing how the block will execute.

`timestamp`, `l1Height`,`l1BlockHash`, and `difficulty` can all impact the execution of transactions on L2, and so they also need to be some predictable value even if the block is being proposed at some later L1 slot.